vast-client/auth.go

```diff
@@ -0,0 +1,162 @@
+package vast_client
+
+import (
+	"bytes"
+	"context"
+	"crypto/tls"
+	"encoding/json"
+	"io"
+	"net/http"
+	"net/url"
+	"strconv"
+	"time"
+
+	"github.com/hashicorp/terraform-plugin-log/tflog"
+)
+
+// TOKEN_REFRESH_TIME_IN_SECONDS Time duration is set to 10 min after this we refresh the token
+const TOKEN_REFRESH_TIME_IN_SECONDS time.Duration = time.Duration(time.Minute * 10)
+
+type Authenticator interface {
+	Authorize(ses Session) error
+	SetAuthHeader(ses Session, headers *http.Header) error
+}
+
+func CreateAuthenticator(ctx context.Context, config *RestClientConfig) Authenticator {
+	// Check if username and password are provided
+	if config.Username != "" && config.Password != "" {
+		// Return a new JWTAuthenticator
+		tflog.Debug(ctx, "VMS session is using username/password authentication (Bearer token will be acquired).")
+		return &JWTAuthenticator{
+			Username: config.Username,
+			Password: config.Password,
+			Token:    nil, // Initially no token
+		}
+	}
+	// If apiToken is provided, return a new ApiRTokenAuthenticator
+	if config.ApiToken != "" {
+		tflog.Debug(ctx, "VMS session is using API token authentication.")
+		return &ApiRTokenAuthenticator{
+			Token: config.ApiToken,
+		}
+	}
+	// If neither are provided, panic with an error message
+	panic("CreateAuthenticator: neither username/password nor apiToken are provided")
+}
+
+type jwtToken struct {
+	Access  string `json:"access"`
+	Refresh string `json:"refresh"`
+	Created time.Time
```

Comment:
```suggestion
	CreatedAt time.Time
```
Maybe like this? `Created` indicates it's a boolean. Or is it me coming from Python? 😁

Reply:
Done